dali/pipeline/operators/geometric/flip.cc

```diff
@@ -0,0 +1,130 @@
+                1)
+        .AddOptionalArg("vertical",
+                R"code(The color space of the output image)code",
+                0);
+
+template <>
+Flip<CPUBackend>::Flip(const OpSpec& spec)
+    : Operator<CPUBackend>(spec)
+    , _horizontal(spec.GetArgument<int32>("horizontal"))
+    , _vertical(spec.GetArgument<int32>("vertical")) {}
+
+
+enum class FlipAxis {
+  HORIZONTAL,
+  VERTICAL,
+  BOTH
+};
+
+template <FlipAxis D>
+struct FlipImpl {};
+
+template <>
+struct FlipImpl<FlipAxis::HORIZONTAL> {
+  template <typename T>
+  static void run(T *output, const T *input, size_t height, size_t width, size_t channels) {
+    for (size_t r = 0; r < height; ++r) {
+      for (size_t c = 0; c < width; ++c) {
+        std::copy(
+            &input[(r * width + c) * channels],
+            &input[(r * width + c + 1) * channels],
+            &output[(r * width + width - c - 1) * channels]);
+      }
+    }
+  }
+};
+
+template <>
+struct FlipImpl<FlipAxis::VERTICAL> {
+  template <typename T>
+  static void run(T *output, const T *input, size_t height, size_t width, size_t channels) {
+    for (size_t r = 0; r < height; ++r) {
+      std::copy(
+          &input[r * width * channels],
+          &input[(r + 1) * width * channels],
+          &output[(height - r - 1) * width * channels]);
+    }
+  }
+};
+
+template <>
+struct FlipImpl<FlipAxis::BOTH> {
+  template <typename T>
+  static void run(T *output, const T *input, size_t height, size_t width, size_t channels) {
+    for (size_t r = 0; r < height; ++r) {
+      for (size_t c = 0; c < width; ++c) {
+        std::copy(
+            &input[(r * width + c) * channels],
+            &input[(r * width + c + 1) * channels],
+            &output[((height - r - 1) * width + width - c - 1) * channels]);
+      }
+    }
+  }
+};
+
+template <FlipAxis D>
+void RunFlip(Tensor<CPUBackend> &output, const Tensor<CPUBackend> &input) {
+  DALI_TYPE_SWITCH(input.type().id(), DType,
+    auto output_ptr = output.mutable_data<DType>();
+    auto input_ptr = input.data<DType>();
+    if (input.GetLayout() == DALI_NHWC) {
+      ssize_t height = input.dim(0), width = input.dim(1), channels = input.dim(2);
+      FlipImpl<D>::run(output_ptr, input_ptr, height, width, channels);
+    } else if (input.GetLayout() == DALI_NCHW) {
+      ssize_t height = input.dim(1), width = input.dim(2), channels = input.dim(0);
+      for (ssize_t c = 0; c < channels; ++c) {
+        auto slice_origin = c * height * width;
+        FlipImpl<D>::run(output_ptr + slice_origin, input_ptr + slice_origin, height, width, 1);
+      }
+    }
+  )
+}
+
+template <>
+void Flip<CPUBackend>::RunImpl(Workspace<CPUBackend>* ws, const int idx) {
+  const auto &input = ws->Input<CPUBackend>(idx);
+  auto &output = ws->Output<CPUBackend>(idx);
+  output.SetLayout(input.GetLayout());
+  output.set_type(input.type());
+  output.ResizeLike(input);
+  DALI_ENFORCE(input.ndim() == 3);
+  if (_vertical && _horizontal) {
```

Comment:
Following up my comment before, I believe we need _vertical and _horizontal to be defined for each sample